README.md
# GrumpTech Garage Door
----------------------
A plug-in component for [Homebridge](https://github.com/nfarina/homebridge "GitHub link") that provides control and status of one or more garage doors using a Raspberry PI. _(some assembly required)_

To provide some measure of security, preventing unintended opening/closing of the doors, a _software_ lock will prevent control of the door. The lock is enabled by default and the state of the lock is not preserved between sessions. (Refer to the _Limitations and Known Issues_ section below)

_GrumpTech Garage Door_ supports both a proximity switch and [sonar based](https://lastminuteengineers.com/arduino-sr04-ultrasonic-sensor-tutorial/ "HC-SR04") sensors to detect the state of the door.

## Installation
This module is not intended to provide an extensible _API_.

> npm i homebridge-grumptechgaragedoor

##  Configuration
The plug-in allows for user specified configuration of the digital input/output lines to use as well as the type of sensors to use for detecting the door. The *config_sample.json* file located in *./config* shows an example of many of the user configurable settings.

### Bridge<br />(bridge)
Key | Description | Value(s) | Default(s) | Remark(s)
:--- | :----------- | :-------- | :---------- | :---------
`username` | Identifier in the form of a globally-unique colon-delimited EUI-48 multicast [MAC](https://en.wikipedia.org/wiki/MAC_address "MAC Address") address. | MAC Address | | Takes the form (xx:xx:xx:xx:xx:xx)

### Platforms<br />(platforms)
Settings for the GrumpTech Garage Door platform.

Key | Description | Value(s) | Default(s) | Remark(s)
:--- | :----------- | :-------- | :---------- | :---------
`platform` | Identifier for the name of the Homebridge platform.  | Any valid non-null string | GrumpTechGarageSystemPlatform |Must match the text in the *config_info/platform* entry of the package.json file.<br /><br />*Recommended to leave as-is.<br/>Rebuild required if changed.*
`name` | Identifier for the name of the Homebridge plug-in. | Any valid non-null string | homebridge-GrumpTechGarageSystem | Identifier for the name of the Homebridge plug-in. Must match the text in the *config_info/plugin* entry of the package.json file.<br /><br/>*Recommended to leave as-is.<br/>Rebuild required if changed.*

#### System<br />(platforms/platform/system)
Settings used to define & configure the garage door system.

Key | Description | Value(s) | Default(s) | Remark(s)
:--- | :----------- | :-------- | :---------- | :---------
`gpio_mode` | *(Optional)* Specify the pin numbering system to use for identifying & accessing the digital input/output resources on the Raspberry Pi. | BCM, RPI | BCM |Refer to the [rpi-gpio](https://github.com/JamesBarwell/rpi-gpio.js "rpi-gpio") node module for more documentation.<br /><br />*Note: Most development and testing was done using BCM mode*
`heartbeat` | *(Optional)* The digital output used to toggle a heartbeat. | Any valid digital output resource on the RPi | 4 (BCM) / 7 (RPI) | Assumed to be connected to a LED or other indicator.<br /><br />*Must be specified according to the 'gpio_mode'.*

#### Doors<br />(platforms/platform/system/doors)
An array of doors in the system.  

Key | Description | Value(s) | Default(s) | Remark(s)
:--- | :----------- | :-------- | :---------- | :---------
`name` | Identifier for the door | Any valid non-null string | | Door names must be unique, as they are used internally for identification.
`state_indicator` | The digital output used to show the state of the door | Any valid digital output resource on the RPi | | Assumed to be connected to a LED or other indicator.<br /><br />*Must be specified according to the 'gpio_mode'.*
`control_request` | The digital output used to control a relay that is used to initiate the door open/close operation | Any valid digital output resource on the RPi | | *Must be specified according to the 'gpio_mode'.*
`manual_control_reqest` | The digital input that is used to initiate a door open/close operation from direct hardware access. | Any valid digital input resource on the RPi | | Assumed to be a push button or some other form of digital input control.<br /><br />*Must be specified according to the 'gpio_mode'.*
`soft_locked` | The initial state of the software lock. | true, false | true | _(Optional)_<br /><br />_true_ indicates that the software lock default to the locked state.<br /><br />Refer to the _Limitations and Known Issues_ section

#### Detection Sensors<br />(platforms/platform/system/doors/detect_sensors)
An array of sensors used to detect the state of the door. The plug-in only functionally supports two active sensors. One sensor to detect the door opening and another to detect closing. Additional sensors can be specified even though not active, which can be helpful during development & debug.

Key | Description | Value(s) | Default(s) | Remark(s)
:--- | :----------- | :-------- | :---------- | :---------
`id` | *Unique* name identifying this sensor | Any non-null string | | The *id* field is only used for debugging.<br />Uniqueness of the name is not checked or enforced
`class` | Classification/Type for this sensor | SonarSensor, ProximitySwitchSensor | | Case sensitive
`function` | Detection function for this sensor | OPEN, CLOSE | | Only the first _OPEN_ and _CLOSE_ sensor will be detected. A function other than _OPEN_ or _CLOSE_ will be ignored.<br /><br /> Case Insensitive.

#### Sensor Configuration<br />(platforms/platform/system/doors/detect_sensors/config)
A collection of configuration settings appropriate for the sensor classification specified.  

Sonar Sensors  

Key | Description | Value(s) | Default(s) | Remark(s)
:--- | :----------- | :-------- | :---------- | :---------
`polling_interval` | *(Optional)* Interval, specified in seconds, over which to measure distance from the sensor | Any positive number greater than or equal to 0.25 seconds. | 5.0 seconds | The minimum bound accounts for the maximum measurement cycle of 60ms-100ms and prevents overlapping trigger requests with the resulting echo signal
`trigger_out` | The digital output used to initiate a sonar distance measurement. | Any valid digital output resource on the RPi. | | *Must be specified according to the 'gpio_mode'.*
`echo_in` | The digital input that is used read the *echo* signal from the sensor, which is used to determine the distance to the object being measured. | Any valid digital input resource on the RPi | | *Must be specified according to the 'gpio_mode'.*
`distance_threshold_change_notification` | *(Optional)* Distance threshold, specified in meters, which must be exceeded for the sensor to raise a *distance_changed* event. | Any positive number greater than 0.0 | 0.08 meters | The *distance_changed* event is not currently used in this application |
`detect_threshold_min` | Minimum threshold, specified in meters, to define a measurement range for detecting the object. | Any number greater than or equal to 0.0 | | Used in conjunction with `detect_threshold_max`. Distance measurements outside of the defined range will result in the sensor indicating that the object is *not detected*<br /><br />Must be less than `detect_threshold_max`
`detect_threshold_max` | Maximum threshold, specified in meters, to define a measurement range for detecting the object. | Any number greater than 0.0 | | Used in conjunction with `detect_threshold_min`. Distance measurements outside of the defined range will result in the sensor indicating that the object is *not detected*<br /><br />Must be greater than `detect_threshold_min`

Proximity Sensor  

Key | Description | Value(s) | Default(s) | Remark(s)
:--- | :----------- | :-------- | :---------- | :---------
`detect_in` | The digital input that is used to detect the door state, as defined by the sensor | Any valid digital input resource on the RPi | | *Must be specified according to the 'gpio_mode'.*
`debounce_time` | *(Optional)* The time, in seconds, to debounce the sensor input | Any number greater than or equal to 1.0 | 1.0 *second* | This is the time required to see no further signal change after detecting the most recent signal state change
`mode` | *(Optional)* Flag indicating if the switch is configured as _normally closed_ or _normally open_ | true, false | true | _true_ indicates _normally closed_  and _false_ indicates _normally open_

## Limitations and Known Issues
Some applications may not properly render these accessories by limiting the accessory characteristics presented to the user. Notably, the default Apple Home application for iOS and macOS is among these. At the present time, it is recommended that the Home+ application for iOS by Matthias Hochgatterer (https://apps.apple.com/us/app/home-4/id995994352) be used to control these accessories.  
* Software Lock: The Apple Home application for both iOS and macOS does not present the user with the ability to control the lock. To work around this problem, the user may configure the plugin to set the default lock state to _unlocked_. However, the recommended solution is to use the Home+ application to control the plugin.  


CHANGELOG.md
# GrumpTech Garage Door Release History
----------------------

Version | Release Date | Comments
:------ | :----------: | :-------
1.0.3   | 2020-MAR-14  | Initial release to npm
1.0.4<br/>1.0.5 | 2020-MAY-10  | * Bug#11: Allow software lock to default to the UNSECURED state.<br>* Add npm build script and repository cleanup
1.0.6   | 2020-MAY-10  | Bug#12: Better handling of attempts to open/close the door when locked.


src/sensorBase.js
/* ==========================================================================
   File:               sensorBase.js
   Class:              SensorBase
   Description:	       Provide common status and control of detection sensors.
   Copyright:          Jun 2019
   ========================================================================== */
'use strict';

// External dependencies and imports.
const _debug = require('debug')('sensorBase');
import { EventEmitter } from 'events';

/* Enumeration for Sonar Results */
const SENSOR_RESULT = {
  UNKNOWN    : 'UNKNOWN',
  UNDETECTED : 'UNDETECTED',
  DETECTED   : 'DETECTED'
};

/* SensorBase represents a base class of detection sensors.

   @event 'result_changed'  => function(oldResult, newResult, context) {}
          Emitted when the sensor detects a state change.
          oldResult:   the prior SENSOR_RESULT
          newResult:   the new SENSOR_RESULT
          context:    reference to the instance of the object raising the event.
*/
class SensorBase extends EventEmitter {
  /* ========================================================================
     Description: Constructor for an instance of a detection sensor.

     Parameters:  identifier: Identifier to associate with this object.
     Return:      N/A
     ======================================================================== */
  constructor(identifier) {

    // Initialize the base class.
    super();

    // Assign the identifier.
    this._id = identifier;

    this._initialized = false;

    // Initialize the result
    this._result = SENSOR_RESULT.UNKNOWN;
  }

  /* ========================================================================
     Description: Property for Sonar Sensor Result

     Parameters:  None

     Return:      Last know SENSOR_RESULT
     ======================================================================== */
  get Result() {
    return (this._result);
  }

  /* ========================================================================
     Description: Read-Only Property for the identifier of this sensor

     Parameters:  None

     Return:      Identifier of this object
     ======================================================================== */
  get Identifier() {
    return this._id;
  }

  /* ========================================================================
     Description: Read-Only Property for the initialized state of this sensor

     Parameters:  None

     Return:      Initialized of this object
     ======================================================================== */
  get Initialized() {
    return this._initialized;
  }

  /* ========================================================================
     Description: Internal helper to update the cached result and notify
                  registered clients of the 'results_changed' event.

     Parameters:  newResult: The updated Sensor Result.

     Return:      None.

     Remarks:     Intended to only be invoked from objects deriving from this class.
     ======================================================================== */
  _setResult(newResult) {

    const oldResult = this._result;

    if (oldResult != newResult) {
      // Update the cached result.
      this._result = newResult;

      // Notify clients - asynchronously.
      setTimeout((caller, resultOld, resultNew) => {
        caller.emit('result_changed', resultOld, resultNew, caller);
      }, 0, this, oldResult, newResult);
    }
  }
}

export {SensorBase as default, SENSOR_RESULT};


src/sonarSensor.js
/* ==========================================================================
   File:               sonarSensor.js
   Class:              SonarSensor
   Description:	       Provide status and control of a sonar sensor used for
                       detection.
   Copyright:          Jun 2019
   ========================================================================== */
'use strict';

// External dependencies and imports.
const _debug  = require('debug')('sonarSensor');
const _gpio   = require('rpi-gpio');
const _gpiop  = _gpio.promise;

// Internal dependencies
import _sensorBase, * as modSensorBase from './sensorBase.js';

// Enumeration for managing the current sonar signal.
const _SONAR_STATE = {
  INACTIVE  : 'INACTIVE',
  ARMED     : 'ARMED',    /* Waiting for rising edge  */
  PENDING   : 'PENDING'   /* Waiting for falling edge */
};

/* The nominal speed of sound in air @20 degC */
const _NOMINAL_SPEED_OF_SOUND                     = 343.0/*meters per sec*/;
/* Minimum polling interval of the sonar sensor. Accounts for the 100ms maximum detection interval of the HS204 */
const _MINIMUM_SONAR_POLLING_INTERVAL             =  0.25/*sec*/;
/* Default polling interval of the sonar sensor */
const _DEFAULT_SONAR_POLLING_INTERVAL             =  5.0/*sec*/;
/* Change threshold to determine when the 'distance_changed' event should be raised. */
const _DISTANCE_THRESHOLD_FOR_CHANGE_NOTIFICATION = 0.08/*meters*/;
/* Time for toggling the trigger to initiate a sonar reading. */
const _SONAR_TRIGGER_TIME                         =  1/*ms*/;
/* Distance used to indicate an invalid distance measurement. */
const _INVALID_DISTANCE                           = -1000.0/*meters*/;


/* SonarSenaor represents a HS204 Sonar Sensor. The sensor is monitored,
   and controlled by a RaspberryPi.

   @event 'distance_changed' => function(oldDistance, newDistance, context) {}
          Emitted when the sonar sensor detects a distance change that exceeds a change threshold.
          oldDistance: the prior distance measured in meters.
          newDistance: the updated distance measured in meters.
          context:     reference to the instance of the object raising the event.
*/
class SonarSensor extends _sensorBase {
  /* ========================================================================
     Description: Constructor for an instance of a sonar sensor.

     Parameters:  identifier:                                 Identifier to associate with this object.
                  configuration:                              Object with the following fields.
                  { polling_interval:                         (Optional) Time in seconds that the sensor should be polled for a distance measurement.
                                                              Default: _DEFAULT_SONAR_POLLING_INTERVAL
                    distance_threshold_change_notification:   (Optional) Distance in meters to use as a change threshold for raising a distance_changed event.
                    trigger_out:                              GPIO Output Channel for the trigger to activate the sonar sensor.
                                                              ** Assumed to be in the GPIO Mode as the provider of the configuration.
                    echo_in:                                  GPIO Input Channel for to measure the door sonar sensor.
                                                              ** Assumed to be in the GPIO Mode as the provider of the configuration.
                    detect_threshold_min                      Minumum distance in meters to qualify as 'detected'.
                    detect_threshold_max                      Maximum distance in meters to qualify as 'detected'.
                  }
     Return:      N/A
     ======================================================================== */
  constructor(identifier, configuration) {

    // Initialize the base class.
    super(identifier);

    _debug(`Constructing sonar sensor`);

    // Verify that the configuration is valid. Throws an exception if false.
    SonarSensor.ValidateConfiguration(configuration);

    // Consume configuration
    /* Polling interval is optional. */
    this._sonarPingInterval = (configuration.hasOwnProperty('polling_interval') ? configuration.polling_interval : _DEFAULT_SONAR_POLLING_INTERVAL) * 1000.0/* milliseconds / second */;
    /* Distance threshold change notification is optional */
    this._distanceThreshold = (configuration.hasOwnProperty('distance_threshold_change_notification') ? configuration.distance_threshold_change_notification : _DISTANCE_THRESHOLD_FOR_CHANGE_NOTIFICATION);
    this._gpioTriggerOut    = configuration.trigger_out;
    this._gpioEchoIn        = configuration.echo_in;
    this._detectMin         = configuration.detect_threshold_min;
    this._detectMax         = configuration.detect_threshold_max;

    // Other data members
    this._initialize();

    /* Create a function pointer for state change notifications. */
    this._myStateChangeCB   = this._stateChange.bind(this);

    _debug(`New Sonar Sensor: id=${this.Identifier}`);
  }

  /* ========================================================================
     Description: Destructor for an instance of a sonar sensor.

     Parameters:  None

     Return:      None

     Remarks:     DO NOT reset the underlying GPIO peripheral. That should only
                  be handled by the garage controller system. Simply make this
                  object inert and set the outputs back to default.
     ======================================================================== */
   async Terminate() {

    _debug(`Terminating Sonar Sensor: ${this.Identifier}`);

    // Stop the sonar.
    this.Active = false;

    // Unregister for change notifications
    _gpio.off( 'change', this._myStateChangeCB );

    super._initialized = false;
  }

  /* ========================================================================
     Description: Start/Initialize the sonar sensor.

     Parameters:  None

     Return:      Flag indicating if the sensor has been initialized.
     ======================================================================== */
  async Start() {

    _debug(`Starting Sonar: ${this.Identifier}`);

    // Re-initialize the internal data.
    this._initialize();

    // Create an array of promises to configure the GPIO and wait for all of them
    // to complete.
    await Promise.all([/* Configure the sonar trigger and initialize to LOW */
                       _gpiop.setup(this._gpioTriggerOut, _gpio.DIR_LOW),
                       /* Configure the door sonar echo input channel to detect all events to allow
                          for the measurement of the pulse duration, which equates to distance. */
                       _gpiop.setup(this._gpioEchoIn,     _gpio.DIR_IN,  _gpio.EDGE_BOTH),
                      ])
    .then(async (setupResults) => {
      // Register for state change events for all input channels.
      _gpio.on( 'change', this._myStateChangeCB );

      // Start the sonar measuring system.
      this.Active = true;

      // Indicate that the sonar is now initialized and ready to operate.
      super._initialized = true;
    })
    .catch((err) => {
      _debug(`Init Error(Sonar:${this.Identifier}): ${err.toString()}`);
    });

    return this.Initialized;
  }

  /* ========================================================================
     Description: Property for Sonar Sensor Activation

     Parameters:  None

     Return:      true if the sonar is active. false otherwise.
     ======================================================================== */
  get Active() {
    let active = false;

    if (this._sonarPingTriggerId) {
      active = true;
    }
    return (active);
  }

  /* ========================================================================
     Description: Property for Sonar Sensor Activation

     Parameters:  active: Flag indicating if the sonar should be active or not.

     Return:      None
     ======================================================================== */
  set Active(active) {
    let reset = !active;
    if (active && !this.Active)
    {
      /* Requesting to activate an inactive sonar. */
      this._sonarPingTriggerId = setInterval( (() => {this._triggerSonar(); }), this._sonarPingInterval);
      // Also reset the data if activating the sonar sensor.
      reset = true;
    }
    else if (!active && this.Active)
    {
      /* Requesting to deactivate an active sonar */
      this._sonarPingTriggerId = clearInterval(this._sonarPingTriggerId);
    }
    else {
      _debug(`Sonar ${this.Identifier}: Sonar is already ${active ? 'active' : 'inactive'}`);
    }

    // Clear the sonar data if inactive or newly activated.
    if (reset)
    {
      this._lastDistanceReading       = _INVALID_DISTANCE;
      this._referenceDistanceReading  = _INVALID_DISTANCE;
      this._referenceTime             = 0;
      this._distanceAge               = 0;
      this._sonarState                = _SONAR_STATE.INACTIVE;
    }
  }

  /* ========================================================================
     Description: Helper to initialize the internal data. Used to support re-starting.

     Parameters:  None

     Return:      None
     ======================================================================== */
  _initialize() {
    super._initialized  = false;
    this.Active         = false;
  }

  /* ========================================================================
     Description: Trigger the door sonar signal to initiate a reading.

     Parameters:  None

     Return:      None
     ======================================================================== */
  _triggerSonar() {
    // The door sonar activates with an active high signal.
    const ACTIVATE_SONAR = true;

    // Ensure that the sonar is not busy.
    if (this._sonarState != _SONAR_STATE.INACTIVE)
    {
      // Trigger is occuring during an existing reading.
      this.Active = false;
      _debug(`Sonar (${this.Identifier}) Unexpected trigger request.`);
    }
    try {
      // Trigger the sonar.
      _gpiop.write(this._gpioTriggerOut, ACTIVATE_SONAR)
      .then(() => {
        // Once active, deactivate the sonar after a brief period.
        return setTimeout(_gpiop.write, _SONAR_TRIGGER_TIME, this._gpioTriggerOut, !ACTIVATE_SONAR);
      })
      .then(() => {
        // Set the Sonar State to armed.
        this._sonarState = _SONAR_STATE.ARMED;
      })
      .catch((error) => {
        this.Active = false;
        _debug(`Sonar (${this.Identifier}) Request Error: ${error.toString()}`);
      });
    }
    catch (err) {
      this.Active = false;
      _debug(`Failed to trigger Sonar (${this.Identifier}): (${err})`);
    }
    finally {
      // Restart the sonar if it deactivated due to an error.
      if (!this.Active) {
        _debug(`Sonar (${this.Identifier}) Restart sonar`);
        this.Active = true;
      }
    }
  }

  /* ========================================================================
     Description: Event handler for State Change events on GPIO Inputs

     Parameters:  channel: GPIO Channel for the event.
                  value:   Value of the GPIO Input

     Return:      None
     ======================================================================== */
  _stateChange(channel, value) {
    switch (channel) {
      case (this._gpioEchoIn) :
      {
        if (this.Active) {
          switch (this._sonarState) {
            case _SONAR_STATE.ARMED :
            {
              if (value) {
                /* The sonar data is Armed. So cache the current time and
                   set the state to PENDING (waiting for final reading) */
                this._referenceTime = Date.now();
                this._sonarState    = _SONAR_STATE.PENDING;
              }
              else {
                _debug(`Sonar (${this.Identifier}): Object too close. Forcing measurement distance of half of the detectable minimum.`);

                /* Perform the distance measurement forcing half the minimum resolution. (JavaScript limitation) */
                this._performDistanceMeasurement(0.5, Date.now());

                // Force the sensor to re-activate.
                this.Active = false;
              }
            }
            break;

            case _SONAR_STATE.PENDING :
            {
              if (!value) {
                /* The sonar data is pending (waiting for the measurement to complete). */
                /* Get the elapsed time while in the armed state. */
                const measurementCompleteTime = Date.now();
                const elapsedTime = measurementCompleteTime - this._referenceTime;

                /* Perform the distance measurement */
                this._performDistanceMeasurement(elapsedTime, measurementCompleteTime);

                /* Set the state back to INACTIVE. */
                this._sonarState = _SONAR_STATE.INACTIVE;
              }
              else {
                _debug(`Sonar (${this.Identifier}): Unexpected GPIO Signal when Pending.`);
                this.Active = false;
              }
            }
            break;

            case _SONAR_STATE.INACTIVE :
            // Break intentionally missing
            default:
            {
              this.Active = false;
            }
            break;
          }

          // Restart if we became active in this cycle.
          if (!this.Active) {
            this.Active = true;
          }
        }
        else {
          _debug(`Sonar (${this.Identifier}): Unexpected GPIO Signal when Not Active.`);
        }
      }
      break;

      default:
      {
        _debug(`Sonar (${this.Identifier}): Unhandled State Change: Chan:${channel} Val:${value}`);
      }
      break;
    }
  }

  /* ========================================================================
     Description: Helper to compute the measured distance and notify clients
                  as needed.

     Parameters:  elapsedTime:              Time elapsed, in milli-seconds, that the distance
                                            measurement was monitored.
                  measurementCompleteTime:  Time of measurement completion, in milli-seconds.

     Return:      None
     ======================================================================== */
  _performDistanceMeasurement(elapsedTime, measurementCompleteTime) {
    /* Distance factor: The sonar must travel to the target and return.
     Therefore the total distance traveled is 2x the distance to the target. */
    const DISTANCE_FACTOR = 2.0;

    /* Convert the elapsed time to distance and cache the time of the measurement. */
    this._lastDistanceReading = (_NOMINAL_SPEED_OF_SOUND * (elapsedTime / 1000.0/* sec / ms */)) / DISTANCE_FACTOR;
    this._distanceAge         = measurementCompleteTime;

    // Raise the distance_changed event, if appropriate.
    if (Math.abs(this._lastDistanceReading - this._referenceDistanceReading) >= this._distanceThreshold) {
      // Raise the event, asynchronously.
      setTimeout((caller, resultOld, resultNew) => {
        caller.emit('distance_changed', resultOld, resultNew, caller);
      }, 0, this, this._referenceDistanceReading, this._lastDistanceReading);

      /* Update the reference distance */
      this._referenceDistanceReading = this._lastDistanceReading;
    }

    // Compute the sensor state.
    const newResult = (((this._lastDistanceReading >= this._detectMin) &&
                        (this._lastDistanceReading <= this._detectMax)) ? modSensorBase.SENSOR_RESULT.DETECTED : modSensorBase.SENSOR_RESULT.UNDETECTED);
    _debug(`Sonar (${this.Identifier}): Measured Distance=${this._lastDistanceReading} Result=${newResult}`);

    // Update the result & raise the result_changed event, if appropriate.
    super._setResult(newResult);
  }

  /* ========================================================================
   Description:    Validate the configuration for the sonar sensor

   Parameters:     configuration: Homebridge configuration node for a sonar sensor

   Return:         true if configuration is valid

   Remarks:        Static method to allow configuration to be validated without creating an object instance
   ======================================================================== */
  static ValidateConfiguration(configuration) {
    const configValid = ( (!configuration.hasOwnProperty('polling_interval')                        || ((typeof(configuration.polling_interval)                        === 'number')  && (configuration.polling_interval >= _MINIMUM_SONAR_POLLING_INTERVAL)))  && /* Optional configuration item */
                          (!configuration.hasOwnProperty('distance_threshold_change_notification')  || ((typeof(configuration.distance_threshold_change_notification)  === 'number')  && (configuration.distance_threshold_change_notification > 0.0)))         && /* Optional configuration item */
                          ( configuration.hasOwnProperty('trigger_out')                             &&  (typeof(configuration.trigger_out)                             === 'number'))                                                                           &&
                          ( configuration.hasOwnProperty('echo_in')                                 &&  (typeof(configuration.echo_in)                                 === 'number'))                                                                           &&
                          ( configuration.hasOwnProperty('detect_threshold_min')                    &&  (typeof(configuration.detect_threshold_min)                    === 'number')  && (configuration.detect_threshold_min >= 0.0))                           &&
                          ( configuration.hasOwnProperty('detect_threshold_max')                    &&  (typeof(configuration.detect_threshold_max)                    === 'number')  && (configuration.detect_threshold_max >  0.0))                           &&
                          (configuration.detect_threshold_max > configuration.detect_threshold_min)                                                                                                                                                               );

    if (!configValid) {
      _debug(`SonarSensor invalid configuration. configuration:${JSON.stringify(configuration)}`);
    }

    return configValid;
  }
}

export default SonarSensor;
